### ISSUE TITLE
Escaping special character

### ISSUE THREAD

**Issue by 107dipan** [vespa-engine/vespa#25105]
I am trying to query vespa with special chars like "+". I am referring to this page and if my understanding is correct we dont need to escape chars for ascii between 32 and 126. Since ascii of + falls between this we should be able to search without escaping special charater
yql - select * from sources * where defaultField_s contains "+". 
The document contains field - defaultField : [ "+" , "another field"]
. https://docs.vespa.ai/en/reference/document-select-language.html under "Locale / Character sets"

Can someone help me with this issue

**Comment by kkraune** [comment]
Please see https://docs.vespa.ai/en/query-rewriting.html#special-tokens - "+" is not a text character

**Comment by bratseth** [comment]
Yes, only characters classified as text are indexed if you use default linguistics processing and default word/text matching.

Are you doing a query, or writing a document selection to select a subset of your data for streaming (visiting)? Document selection is only relevant for the latter.

If you want to _query_ for "+", you can either switch to exact matching (suitable for short fields you'll match completely/by prefix and probably store as attributes), or otherwise plug in your own linguistics component to do tokenization the way you want.

**Comment by bratseth** [comment]
The default tokenization (for full text indexes) will match those, since they will become the token lists ["2", "2"] and ["harry", "ron"] on both the document and query side. They will also match these documents when queried for just "2", or "ron" etc., but ranking will be able to sort more complete matches higher.

**Comment by 107dipan** [comment]
Thanks a lot Jon! 
If we enable match "exact" on the text fields would that have any effect on the ingestion or search latency?

**Comment by bratseth** [comment]
Are these short field string values, not full text fields? If so you can enable exact match. It will not have any impact on latency.

**Comment by 107dipan** [comment]
These are short field string values. I converted the field type to exact match but I am still having issue forming the query
select * from sources * where table contains "harry&ron". I also tried to escape the charater

**Comment by bratseth** [comment]
So, what query do you see it produce? Set tracelevel=1.

**Comment by 107dipan** [comment]
I am getting the following error message 
"message": "Could not create query from YQL: query:L1:54 no viable alternative at input 'defaultField_s contains \"'",
                "stackTrace": "com.yahoo.processing.IllegalInputException: com.yahoo.search.yql.ProgramCompileException: query:L1:54 no viable alternative at input 'defaultField_s contains \"'\n\tat com.yahoo.search.yql.YqlParser.parseYqlProgram(YqlParser.java:809)

**Comment by bratseth** [comment]
A query parse error. Could you share a full YQL query which produces this?

**Comment by 107dipan** [comment]
http://<container-endpoint>/search/?yql=select * from sources * where (defaultField_s matches "harry&ron")

**Comment by bratseth** [comment]
> Do u mean url-escape the char "&"

Yes, and the quotes. Just run all strings through an url-encoder.

Special tokens are for full-text search, so not relevant here I think.

**Comment by 107dipan** [comment]
By full-text search did u mean using "exact" match mode.

**Comment by bratseth** [comment]
No I mean what you need for fields that contain arbitrarily long general text data, and therefore need token-based (partial) matching and full-text relevance, what you get by writing "indexing: index" on a string field.

**Comment by kkraune** [comment]
We have added an example at the end of https://docs.vespa.ai/en/query-rewriting.html for using the vespa cli to create an escaped YQL string as well as an exact match example

**Comment by 107dipan** [comment]
Thanks a lot Kristain and John!
This was very helpful

**Comment by 107dipan** [comment]
Hi, I  was trying to create a query by  escaping wildcards. 
Example - I want to do a wildcard search on attribute fields but we will need to escape some of the "." in the query string. 
Yql - select * from sources *  where attr_field matches \"^wild.card.File.\". I want to escape the 1st and second "." in the string and only the last ">" should be treated as a regular expression. 
I have tried escaping the character by using "\" and also url encoding the string.

**Comment by kkraune** [comment]
Using Query tracing, you will see what the query is parsed to: https://docs.vespa.ai/en/query-api.html#query-tracing. I will add an escaping example to the documentation:
```
#
# The artist field is:
#   "artist": "Meta..ica"
#
$ vespa query -v 'select * from music where artist matches "M.ta\\.\\.ica"'
```
The above query matches, as the first `.` matches any character, and the two latter are literal matches using an escape before each `.`

Hope this helps

**Comment by bratseth** [comment]
There's already https://docs.vespa.ai/en/faq.html#how-does-backslash-escapes-work
